### PR TITLE
Transport restructure

### PR DESCRIPTION
**Pull request by boks1971**
- Attempt to simplify/stream line transport
- Introduce a channel which processes events
- All things async
- When there is any error in the negotiation path, error out and force a reconnect

Reasons for doing this:
- The code was getting a bit convoluted, especially the bits which tried to treat offerer/answerer in the same routine and was getting confusing (at least for me :-) )
- This is a critical piece and as we build more, this getting error prone is a possibility
- There are paths where errors were not handled. For example, if writing response errors out, that would not be tracked anywhere.

### CODE REVIEW

File: pkg/rtc/transport.go [hunk 1]
```diff
@@ -876,18 +779,31 @@ func (t *PCTransport) AddNegotiationPending(publisherID livekit.ParticipantID) {
...
 		})
-		if err := t.CreateAndSendOffer(nil); err != nil {
-			t.params.Logger.Errorw("could not negotiate", err)
-		}
-	} else {
-		t.debouncedNegotiate(func() {
-			if err := t.CreateAndSendOffer(nil); err != nil {
-				t.params.Logger.Errorw("could not negotiate", err)
-			}
+		t.debouncePending = false
```
Comment (boks1971) [inline pkg/rtc/transport.go:786]: Introducing this variable as each unforced negotiate restarts the debounce and pushes things back by 150ms.

File: pkg/rtc/transport.go [hunk 2]
```diff
@@ -84,31 +140,38 @@ type PCTransport struct {
...
-
+	preferTCP atomic.Bool
+	isClosed  atomic.Bool
+
+	eventChMu sync.RWMutex
+	eventCh   chan event
+
+	// the following should be accessed only in event processing go routine
+	cacheLocalCandidates      bool
+	cachedLocalCandidates     []*webrtc.ICECandidate
```
Comment (boks1971) [inline pkg/rtc/transport.go:167]: These should not be required anymore as everything goes through the channel and gets processed in a go routine. But, leaving it in here just for safety.

File: pkg/rtc/transport.go [hunk 3]
```diff
@@ -981,138 +897,312 @@ func (t *PCTransport) configureReceiverDTX(enableDTX bool) {
...
+}
+
+func (t *PCTransport) processEvents() {
+	for event := range t.eventCh {
+		err := t.handleEvent(&event)
+		if err != nil {
+			t.params.Logger.Errorw("error handling event", err, "event", event.String())
+			if t.onNegotiationFailed != nil {
+				t.onNegotiationFailed()
+			}
```
Comment (boks1971) [inline pkg/rtc/transport.go:963]: Just consolidating all processing here. Basically, any error is not recoverable (or at least will lead to confusing state). So, checking this one place and forcing a full reconnect on any error.

Comment (cnderrauber) [reply pkg/rtc/transport.go:1170]: Wondering not all error is critical, if a ice candidate is not valid but others are valid, peerconnection might still connect.

File: pkg/rtc/transport.go [hunk 4]
```diff
@@ -1361,3 +1370,328 @@ func (t *PCTransport) filterCandidates(sd webrtc.SessionDescription) webrtc.Sess
...
+			if onOffer := t.getOnOffer(); onOffer != nil {
+				err := onOffer(*offer)
+				if err != nil {
+					prometheus.ServiceOperationCounter.WithLabelValues("offer", "error", "write_message").Add(1)
+				} else {
+					prometheus.ServiceOperationCounter.WithLabelValues("offer", "success", "").Add(1)
+				}
+				return err
+			}
+			return ErrNoOfferHandler
```
Comment (cnderrauber) [inline pkg/rtc/transport.go:1681]: Add a `else` here?

File: pkg/rtc/transport.go (hunk 3 above)
Comment (boks1971) [reply pkg/rtc/transport.go:1170]: Good point @cnderrauber . But, I was thinking (actually more hoping :-) ) that in those there is no error returned by Pion for `AddICECandidate`. It just ignores it. Will check on it.

Comment (boks1971) [reply pkg/rtc/transport.go:1164]: @cnderrauber Checked this. Pion filters out all non-critical errors and returns nil. So, treating all errors as bad should be fine.

File: pkg/rtc/transport.go [hunk 5]
```diff
@@ -1283,46 +1151,216 @@ func (t *PCTransport) initPCWithPreviousAnswer(previousAnswer webrtc.SessionDesc
...
+	if t.isClosed.Load() {
+		t.eventChMu.RUnlock()
 		return
 	}
 
-	t.streamAllocator.OnStreamStateChange(f)
+	select {
+	case t.eventCh <- event:
+	default:
+		t.params.Logger.Warnw("event queue full", nil)
```
Comment (cnderrauber) [inline pkg/rtc/transport.go:1181]: can add a event field to log

**Review by cnderrauber — APPROVED** [review]

File: pkg/rtc/transport.go (hunk 5 above)
Comment (boks1971) [reply pkg/rtc/transport.go:1181]: Will do

**Pull request merged**